File: app/src/main/java/com/raival/quicktools/common/BackgroundTask.java
package com.raival.quicktools.common;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.raival.quicktools.R;

public class BackgroundTask {
    Runnable preTask, task, postTask;
    AlertDialog alertDialog;

    public BackgroundTask() {

    }

    public void setTasks(Runnable preTask, Runnable task, Runnable postTask) {
        this.preTask = preTask;
        this.task = task;
        this.postTask = postTask;
    }

    public BackgroundTask run() {
        new Thread(() -> {
            new Handler(Looper.getMainLooper()).post(preTask);
            task.run();
            new Handler(Looper.getMainLooper()).post(postTask);
        }).start();
        return this;
    }

    public void dismiss() {
        if (alertDialog != null)
            alertDialog.dismiss();
    }

    @SuppressLint("ResourceType")
    public void showProgressDialog(String msg, Activity activity) {
        alertDialog = new MaterialAlertDialogBuilder(activity)
                .setCancelable(false)
                .setView(getProgressView(msg, activity))
                .show();
    }

    private View getProgressView(String msg, Activity activity) {
        View v = activity.getLayoutInflater().inflate(R.layout.progress_view, null);
        ((TextView) v.findViewById(R.id.msg)).setText(msg);
        return v;
    }
}

File: app/src/main/java/com/raival/fileexplorer/common/BackgroundTask.kt
package com.raival.fileexplorer.common

import android.annotation.SuppressLint
import android.app.Activity
import android.os.Handler
import android.os.Looper
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.raival.fileexplorer.R
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class BackgroundTask {
    private lateinit var preTask: Runnable
    private lateinit var task: Runnable
    private lateinit var postTask: Runnable
    private lateinit var alertDialog: AlertDialog

    fun setTasks(preTask: Runnable, task: Runnable, postTask: Runnable) {
        this.preTask = preTask
        this.task = task
        this.postTask = postTask
    }

    fun run() {
        CoroutineScope(Dispatchers.IO).launch {
            Handler(Looper.getMainLooper()).post(preTask)
            task.run()
            Handler(Looper.getMainLooper()).post(postTask)
        }.start()
    }

    fun dismiss() {
        if (this::alertDialog.isInitialized) alertDialog.dismiss()
    }

    @SuppressLint("ResourceType")
    fun showProgressDialog(msg: String, activity: Activity) {
        alertDialog = MaterialAlertDialogBuilder(activity)
            .setCancelable(false)
            .setView(getProgressView(msg, activity))
            .show()
    }

    private fun getProgressView(msg: String, activity: Activity): View {
        @SuppressLint("InflateParams") val v =
            activity.layoutInflater.inflate(R.layout.progress_view, null)
        (v.findViewById<View>(R.id.msg) as TextView).text = msg
        return v
    }
}